src/engine/sequencer.js

```diff
@@ -168,109 +167,128 @@ class Sequencer {
         return doneThreads;
     }
 
+    /**
+     * Step the requested thread once. The thread must be at a hat block.
+     * @param {!Thread} thread Thread object to step.
+     */
+    stepHat (thread) {
+        execute(this, thread);
+        if (thread.status === Thread.STATUS_RUNNING) thread.goToNextBlock();
+    }
+
+    /**
+     * Pop the stack as long as it is pointing at null.
+     * @param {!Thread} thread Thread to pop.
+     */
+    unwrapStack (thread) {
+        while (
+            thread.status === Thread.STATUS_RUNNING &&
+            thread.peekStack() === null
+        ) {
+            thread.popStack();
+
+            const stackFrame = thread.peekStackFrame();
+            if (stackFrame === null) {
+                // No more stack to run!
+                this.retireThread(thread);
+            } else if (stackFrame.isLoop) {
+                // The current level of the stack is marked as a loop.
+                // Return to yield for the frame/tick in general.
+                thread.status = Thread.STATUS_YIELD;
+            } else {
+                // Step to the next block.
+                thread.goToNextBlock();
+            }
+        }
+    }
+
     /**
      * Step the requested thread for as long as necessary.
      * @param {!Thread} thread Thread object to step.
      */
     stepThread (thread) {
-        let currentBlockId = thread.peekStack();
-        if (!currentBlockId) {
-            // A "null block" - empty branch.
-            thread.popStack();
```

Comment:
I think these changes need to be rebased off of current develop since this PR is missing the fix from #2126